src/Features/Core/Portable/Diagnostics/DiagnosticService.cs

```diff
@@ -314,17 +322,35 @@ private void AssertIfNull(ImmutableArray<DiagnosticData> diagnostics)
         }
 
         [Conditional("DEBUG")]
-        private void AssertIfNull(DiagnosticData diagnostic)
+        private void AssertIfNull<T>(T obj) where T : class
         {
-            if (diagnostic == null)
+            if (obj == null)
             {
                 Contract.Requires(false, "who returns invalid data?");
             }
         }
 
+        [Conditional("DEBUG")]
+        public void AssertCleanup()
+        {
+            foreach (var idAndData in _map.Values.ToList())
+            {
+                foreach (var data in idAndData.Values.ToList())
+                {
+                    AssertIfNull(data.Workspace);
+                }
+            }
+        }
+
         private struct Data : IEquatable<Data>
         {
-            public readonly Workspace Workspace;
+            private static readonly ConditionalWeakTable<Workspace, WeakReference<Workspace>> s_cache =
+                new ConditionalWeakTable<Workspace, WeakReference<Workspace>>();
+
+            // not all workspace lives whole VS session. 
+            // make sure we not strongly hold onto them.
```

Comment:
I'm stating that this change is very concerning because it indicates that lifetimes of many objects in this part of the system have very unclear lifetimes.  The need to add weak referneces and CWT to try to work around this seems like a very poor way to address this problem.

The following have not been explained yet:

1. Why do Data objects ever outlive the workspace they're associated with.
2. Why do Data objects need to hold onto a workspace at all?
3. Why we need to use weak referneces that hold onto heavy data, instead of just holding onto a lightweight piece of data instead.

> we have CCW all over the places

CWTs should only be used when they're the right solution for a specific problem.  For example, when we have no control over lifetimes and we must associate some piece of data with other objects.  I'm not at all convinced that that's the scenario here, and i would far prefer an approach that prevents this leaking problem from happening in the first place.